mouse.go

```diff
@@ -58,45 +58,62 @@ var mouseEventTypes = map[MouseEventType]string{
 //
 //     ESC [M Cb Cx Cy
 //
+// See: http://www.xfree86.org/current/ctlseqs.html#Mouse%20Tracking
 func parseX10MouseEvent(buf []byte) (m MouseEvent, err error) {
 	if len(buf) != 6 || string(buf[:3]) != "\x1b[M" {
 		return m, errors.New("not an X10 mouse event")
 	}
 
 	e := buf[3] - 32
 
-	switch e {
-	case 35:
-		m.Type = MouseMotion
-	case 64:
-		m.Type = MouseWheelUp
-	case 65:
-		m.Type = MouseWheelDown
-	default:
-		switch e & 3 {
-		case 0:
-			if e&64 != 0 {
-				m.Type = MouseWheelUp
-			} else {
-				m.Type = MouseLeft
-			}
-		case 1:
-			if e&64 != 0 {
-				m.Type = MouseWheelDown
-			} else {
-				m.Type = MouseMiddle
-			}
-		case 2:
+	const (
+		bitShift  = 0b0000_0100
+		bitAlt    = 0b0000_1000
+		bitCtrl   = 0b0001_0000
+		bitButton = 0b0010_0000
+		bitWheel  = 0b0100_0000
+
+		bitsRelease = 0b0000_0011
+
+		bitsMask = 0b0000_0011
+
+		bitsLeft   = 0b0000_0000
+		bitsMiddle = 0b0000_0001
+		bitsRight  = 0b0000_0010
+		bitsMotion = 0b0000_0011
+
+		bitsWheelUp   = 0b0000_0000
+		bitsWheelDown = 0b0000_0001
+	)
```

Comment:
I wonder if it makes sense to move these consts to the package level so that they don't need to be reallocated every time we process a mouse event?

Reply:
Currently it doesn't really matter where to place these constants. It's a good practice to place small constants close to the code that use them.

> so that they don't need to be reallocated every time we process a mouse event?

In Go it works a little bit different:
1. In Go replaces constants with its values on compile time. No matter where you place them.
2. Constants are untyped (which means they don't have size before you assign it with a variable / parameter / field / etc.).
3. (//just imagining//) Even if we place them on top level and Go would you them by reference they would allocate `size_t` (32/64bits) to store pointer :).

---

There is only one reason (in my opinion) to put them on top level if you need them in another function / package.

Reply:
Ah, thanks for the explanation; I didn't consts works like that in Go. In that case it, yeah, it makes sense where you have them. I can't imagine the rest of the package would need them.